src/actions/InsertEmptyLines.ts

```diff
@@ -0,0 +1,102 @@
+import {
+  Action,
+  ActionPreferences,
+  ActionReturnValue,
+  Graph,
+  TypedSelection,
+} from "../Types";
+import { TextEditor, Selection, Position } from "vscode";
+import displayPendingEditDecorations from "../editDisplayUtils";
+import { runForEachEditor } from "../targetUtils";
+
+class InsertEmptyLines implements Action {
+  targetPreferences: ActionPreferences[] = [{ insideOutsideType: "inside" }];
+
+  constructor(
+    private graph: Graph,
+    private insertAbove: boolean,
+    private insertBelow: boolean
+  ) {
+    this.run = this.run.bind(this);
+  }
+
+  async run([targets]: [TypedSelection[]]): Promise<ActionReturnValue> {
+    displayPendingEditDecorations(
+      targets,
+      this.graph.editStyles.referenced,
+      this.graph.editStyles.referencedLine
+    );
+
+    const edits = await runForEachEditor(
+      targets,
+      (target) => target.selection.editor,
+      async (editor: TextEditor, targets: TypedSelection[]) => {
+        const lines = targets.flatMap((target) => {
+          const lines = [];
+          if (this.insertAbove) {
+            lines.push(target.selection.selection.start.line);
+          }
+          if (this.insertBelow) {
+            lines.push(target.selection.selection.end.line + 1);
+          }
+          return lines;
+        });
+        return { editor, lines };
+      }
+    );
+
+    for (const edit of edits) {
+      await edit.editor.edit((editBuilder) => {
+        edit.lines.forEach((line) => {
+          editBuilder.insert(new Position(line, 0), "\n");
+        });
+      });
+    }
+
+    const thatMark = targets.map((target) => {
```

Comment:
So I guess it probably makes sense to have "that" refer to the targets as you've done here, tho it might be useful in the future to support something like a "changed" mark, because I could see wanting to refer to the newly insterted lines as well.

Also, this code can be removed once we have `listenForDocumentChanges`, right?

Reply:
My assumption was that if you wanted the new lines you would use drink and pour instead but I see your point.

Yes I believe so.